File: backend/src/main/java/com/aura_card/backend/controller/SavedCardController.java
package com.aura_card.backend.controller;

import com.aura_card.backend.dto.request.CreateSavedCardRequest;
import com.aura_card.backend.dto.request.UpdateSavedCardRequest;
import com.aura_card.backend.dto.response.ApiResponse;
import com.aura_card.backend.dto.response.SavedCardDetailResponse;
import com.aura_card.backend.dto.response.SavedCardListItemResponse;
import com.aura_card.backend.service.SavedCardService;
import jakarta.validation.Valid;
import jakarta.validation.constraints.Min;
import lombok.RequiredArgsConstructor;
import org.springframework.data.domain.Page;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.security.access.prepost.PreAuthorize;
import org.springframework.security.core.annotation.AuthenticationPrincipal;
import org.springframework.web.bind.annotation.DeleteMapping;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.PathVariable;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.PutMapping;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RequestParam;
import org.springframework.web.bind.annotation.RestController;
import org.springframework.validation.annotation.Validated;

import java.util.List;

@Validated
@RestController
@RequestMapping("/api/saved-cards")
@RequiredArgsConstructor
public class SavedCardController {

        private final SavedCardService savedCardService;

        // Lấy danh sách thiệp lưu của user
        @GetMapping
        @PreAuthorize("hasRole('CUSTOMER')")
        public ResponseEntity<ApiResponse<List<SavedCardListItemResponse>>> getSavedCards(
                        @AuthenticationPrincipal String userId,
                        @RequestParam(defaultValue = "1") @Min(1) int page,
                        @RequestParam(defaultValue = "6") @Min(1) int limit) {

                Page<SavedCardListItemResponse> data = savedCardService.getSavedCards(userId, page, limit);

                return ResponseEntity.ok(ApiResponse.<List<SavedCardListItemResponse>>builder()
                                .message("Lấy danh sách thiệp lưu thành công")
                                .data(data.getContent())
                                .total(data.getTotalElements())
                                .totalPages(data.getTotalPages())
                                .build());
        }

        // Lấy thiệp lưu theo id
        @GetMapping("/{savedCardId}")
        @PreAuthorize("hasRole('CUSTOMER')")
        public ResponseEntity<ApiResponse<SavedCardDetailResponse>> getSavedCardById(
                        @PathVariable String savedCardId) {

                return ResponseEntity.ok(ApiResponse.<SavedCardDetailResponse>builder()
                                .message("Lấy thiệp lưu thành công")
                                .data(savedCardService.getSavedCardById(savedCardId))
                                .build());
        }

        // Thêm thiệp lưu
        @PostMapping
        @PreAuthorize("hasRole('CUSTOMER')")
        public ResponseEntity<ApiResponse<SavedCardDetailResponse>> createSavedCard(
                        @AuthenticationPrincipal String userId,
                        @RequestBody @Valid CreateSavedCardRequest request) {

                return ResponseEntity.status(HttpStatus.CREATED)
                                .body(ApiResponse.<SavedCardDetailResponse>builder()
                                                .message("Lưu thiệp thành công")
                                                .data(savedCardService.createSavedCard(request, userId))
                                                .build());
        }

        // Cập nhật thiệp lưu
        @PutMapping("/{savedCardId}")
        @PreAuthorize("hasRole('CUSTOMER')")
        public ResponseEntity<ApiResponse<SavedCardDetailResponse>> updateSavedCard(
                        @PathVariable String savedCardId,
                        @RequestBody @Valid UpdateSavedCardRequest request) {

                return ResponseEntity.ok(ApiResponse.<SavedCardDetailResponse>builder()
                                .message("Cập nhật thiệp lưu thành công")
                                .data(savedCardService.updateSavedCard(savedCardId, request))
                                .build());
        }

        // Xóa thiệp lưu
        @DeleteMapping("/{savedCardId}")
        @PreAuthorize("hasRole('CUSTOMER')")
        public ResponseEntity<ApiResponse<Void>> deleteSavedCard(@PathVariable String savedCardId) {
                savedCardService.deleteSavedCard(savedCardId);
                return ResponseEntity.ok(ApiResponse.<Void>builder()
                                .message("Xóa thiệp lưu thành công")
                                .build());
        }
}

File: backend-nodejs/controllers/savedcard.controller.js
const savedCardService = require("../services/savedCard.service");
const response = require("../utils/response.util");

// Lấy danh sách thiệp lưu
const getSavedCards = async (req, res, next) => {
  try {
    const { page, limit } = req.query;
    const result = await savedCardService.getSavedCards(req.user.id, {
      page,
      limit,
    });
    response.success(res, {
      message: "Lấy danh sách thiệp lưu thành công",
      data: result.data,
      total: result.total,
      totalPages: result.totalPages,
    });
  } catch (error) {
    next(error);
  }
};

// Lấy thiệp lưu theo id
const getSavedCardById = async (req, res, next) => {
  try {
    const data = await savedCardService.getSavedCardById(
      req.params.savedCardId,
    );
    response.success(res, { message: "Lấy thiệp lưu thành công", data });
  } catch (error) {
    next(error);
  }
};

// Thêm thiệp lưu
const createSavedCard = async (req, res, next) => {
  try {
    const data = await savedCardService.createSavedCard(req.body, req.user.id);
    response.success(res, {
      message: "Lưu thiệp thành công",
      data,
      status: 201,
    });
  } catch (error) {
    next(error);
  }
};

// Cập nhật thiệp lưu
const updateSavedCard = async (req, res, next) => {
  try {
    const data = await savedCardService.updateSavedCard(
      req.params.savedCardId,
      req.body,
    );
    response.success(res, { message: "Cập nhật thiệp lưu thành công", data });
  } catch (error) {
    next(error);
  }
};

// Xóa thiệp lưu
const deleteSavedCard = async (req, res, next) => {
  try {
    await savedCardService.deleteSavedCard(req.params.savedCardId);
    response.success(res, { message: "Xóa thiệp lưu thành công" });
  } catch (error) {
    next(error);
  }
};

module.exports = {
  getSavedCards,
  getSavedCardById,
  createSavedCard,
  updateSavedCard,
  deleteSavedCard,
};
